### PR TITLE
Add subscript operator to Expression to access sub-expressions by name

### PR DESCRIPTION
**Pull request by jteuber**
Hi there!
It might be just me, but accessing certain sub-expressions of a rule by name seems very handy for me, especially with very complicated rules. So I added this. It seems to work quite well. Especially the std::optional makes parsing expressions with optional expressions much simpler.

### CODE REVIEW

**Comment by TheLartians** [comment]
Hey, thanks for the PR, I definitely see the advantages of the rule based query, and good call with the optional return! Could you also add a test case, so the feature remains stable in the future?

**Comment by jteuber** [comment]
Sure, there you go! :)

File: test/source/parser.cpp [hunk 1]
```diff
@@ -284,3 +284,25 @@ TEST_CASE("Documentation Example") {
...
+    return std::string();
+  };
+  program["Start"] << "IntSubscript | StringSubscriptOptional";
+
+  program.setStart(program["Start"]);
+
+  REQUIRE(program.run("ab 0") == "0");
+  REQUIRE(program.run("ab CD 1") == "CD");
+  REQUIRE(program.run("2 ef") == "");
+  REQUIRE(program.run("2 GH ij") == "GH");
```
Comment (TheLartians) [inline test/source/parser.cpp:307]: Hey, thanks for adding the test case! I do find it to be a hard to follow, as it's doing a lot of things at once. What do you think about simplifying it to a single rule that checks the new `[string]` operator?

For example, something like

```cpp
  ParserGenerator<bool> program;
  program["Word"] << "[a-z]+";
  program["Yell"] << "[A-Z]+";
  program["Start"] << "Word | Yell" >> [](auto e){ return bool(e["Yell"]);  }
  REQUIRE(!program.run("hello"));
  REQUIRE(program.run("HELLO"));
```

**Comment by jteuber** [comment]
Yes, you're absolutely right. That's a much better test case.
I had a hard time coming up with a simple test yesterday so I basically simplified the use case I had for the subscript operator.
As soon as I have some time on my PC today I copy over your suggestion.

Feb 5, 2021 10:43:43 Lars Melchior <notifications@github.com>:

**Review by TheLartians — APPROVED** [review]
Awesome, thanks for the PR and updates!

**Comment by TheLartians** [comment]
The feature is released in v2.2.0! 🎉
